### ISSUE TITLE
Numpy Mean gives unencrypted result

### ISSUE THREAD

**Issue by vrush0022** [ibarrond/Pyfhel#57]
I have encrypted every value in a column in a pandas dataframe .
Now when I use np.mean() on that column, it is returning the unencrypted result.
I am expecting it to return the encrypted mean.
Is this an expected behaviour?

**Comment by ibarrond** [comment]
Can you share your code?

I can assure you, if you encrypt all the values, any operations among them will yield encrypted results. Moreover, I believe the `np.mean()` includes a division, which is not supported by Pyfhel yet (an update to CKKS scheme should fix this).

**Comment by vrush0022** [comment]
`import numpy as np
from Pyfhel import Pyfhel, PyPtxt, PyCtxt
from sklearn import datasets
import pandas as pd
from sklearn.model_selection import train_test_split

#Load IRIS Dataset
iris=datasets.load_iris()
data=pd.DataFrame(iris.data,columns=iris.feature_names)

#HE Object Creation
HE = Pyfhel()    
HE.contextGen(p=65537, m=2**12, flagBatching=True) 
HE.keyGen()

#Encrypt each of the columns of original dataframe
enc_data=pd.DataFrame(columns=data.columns)
for col in enc_data.columns:
  enc_data[col]=[HE.encryptFrac(x) for x in data[col].to_numpy()]

#Compute Mean of Sepal Length
np.mean(enc_data['sepal length (cm)'])


`

**Comment by vrush0022** [comment]
![image](https://user-images.githubusercontent.com/43715994/94643641-efa12500-0304-11eb-85cb-97465e5e3c50.png)
Mean returns the unencrypted result whereas sum returns the encrypted result.

**Comment by ibarrond** [comment]
This is weird indeed! It seems like `Pandas.Series` is doing something, but I cannot figure out how can they decrypt the results. 

```python
import numpy as np
from Pyfhel import Pyfhel, PyPtxt, PyCtxt
import pandas as pd

#HE Object Creation
HE = Pyfhel()    
HE.contextGen(p=65537, m=2**12, flagBatching=True) 
HE.keyGen()

#Two float values
a = 1.5
b = 2.5
ca = HE.encryptFrac(a)
cb = HE.encryptFrac(b)
ca_cb = np.array([ca, cb])

# All the following operations raise this exception because the division is not implemented
#   TypeError: unsupported operand type(s) for /: 'Pyfhel.PyCtxt.PyCtxt' and 'int'
(ca+cb)/2
np.mean([ca, cb])
np.mean(np.array([ca, cb]))
np.sum([ca, cb])/2

# This would be a nice way to tackle it, but it isn't implemented yet in Pyfhel
#   TypeError: <Pyfhel ERROR> must be either PyCtxt or PyPtxt
(ca+cb)*0.5

# This is a good way to perform mean
enco_len = HE.encodeFrac(1/len(ca_cb))
encr_sum = np.sum(ca_cb)
encr_mean = encr_sum * enco_len
#> encr_mean 
#>       <Pyfhel.PyCtxt.PyCtxt at 0x7f89413158b8>
#> HE.decryptFrac(encr_mean)
#>       2.0


# This is the weird operation, yielding an unencrypted result 2.0
np.mean(pd.Series(ca_cb))
np.mean(pd.Series([ca, cb]))
pd.Series([ca, cb]).mean()
pd.Series(ca_cb).mean()

**Comment by vrush0022** [comment]
Yes, I did use that way ultimately to calculate the mean. But the biggest concern for me was knowing how can pandas series decrypt the results and is there any scope for such unofficial decryption ?

**Comment by ibarrond** [comment]
Ok, I discovered that somehow `pandas` is decrypting the results when running `mean`.  To do so, I just used a `Pyfhel` instance without a secret key: it should be able to encrypt values and operate on them, but not decrypt them. This is exactly the case:

```python
import numpy as np
from Pyfhel import Pyfhel, PyPtxt, PyCtxt
import pandas as pd


##### CLIENT
#HE Object Creation, including the public and private keys
HE = Pyfhel()    
HE.contextGen(p=65537, m=2**12, flagBatching=True) 
HE.keyGen()

# Saving only the public key and the context
HE.savepublicKey("mypk.pk")
HE.saveContext("mycontx.con")

# Serializing two float values
a = 1.5
b = 2.5
ca = HE.encryptFrac(a)
cb = HE.encryptFrac(b)

ca_str = ca.savem() # We could save/reload these strings in a file
cb_str = cb.savem()

# Computing the mean using `pandas` yields the unencrypted results
pd.Series([ca, cb]).mean()
#> 2.0




##### SEMI-HONEST CLOUD
# Generating a second HE, acting as the honest-but-curious Cloud provider,
#  that will perform the operations and try to decrypt everything
HE_Cl = Pyfhel()    
HE_Cl.restoreContext("mycontx.con")
HE_Cl.restorepublicKey("mypk.pk")

# loading the two ciphertexts. There is clearly potential for improvement here
from Pyfhel.util import ENCODING_t
c2a = PyCtxt()
c2a._pyfhel = HE_Cl
c2a.loadm(ca_str)
c2a._encoding = ENCODING_t.FRACTIONAL

c2b = PyCtxt()
c2b._pyfhel = HE_Cl
c2b.loadm(cb_str)
c2b._encoding = ENCODING_t.FRACTIONAL

# Computing the mean using `pandas` yields the same error as any other attempt
#   to decrypt the results
pd.Series([c2a, c2b]).mean()
HE_Cl.decrypt(c2a)
HE_Cl.decrypt(c2a + c2b)
#> ---------------------------------------------------------------------------
#> RuntimeError                              Traceback (most recent call last)
#> Pyfhel/Pyfhel.pyx in Pyfhel.Pyfhel.Pyfhel.decryptFrac()

#> RuntimeError: Missing a Private Key [...]
```

**Comment by ibarrond** [comment]
TLDR; `pandasSeries(),mean()` is automatically calling `PyCtxt.decryptFrac()`, thus the result is unencrypted. Better stick with `numpy` operations to avoid this unexpected behavior (Pyfhel doesn't implement this explicitely, I wonder how `Pandas` manages to do this).

**Comment by ibarrond** [comment]
I understood what is hapenning! Since Pandas and Numpy cannot operate with `PyCtxt` ciphertexts for division, they both try to cast the ciphertext into `float` or `int`. When they do this, they are efficiently decrypting the ciphertext, as seen here:
https://github.com/ibarrond/Pyfhel/blob/d17712b824a8f1e559755efe2ba96c7c1611bed2/Pyfhel/PyCtxt.pyx#L515-L518

In an example:
```python
from Pyfhel import Pyfhel, PyPtxt, PyCtxt

HE = Pyfhel()    
HE.contextGen(p=65537, m=2**12, flagBatching=True) 
HE.keyGen()

a = 1.5
ca = HE.encryptFrac(a)

print(float(ca))
#> 1.5
```

**Comment by vrush0022** [comment]
Has this issue been fixed?
